packages/dds/tree/src/test/cursor.spec.ts

```diff
@@ -0,0 +1,427 @@
+    suiteName: string,
+    factory: (data: JsonableTree) => ITreeCursor,
+    dataFromCursor: (cursor: ITreeCursor) => JsonableTree,
+): void {
+    describe(`${suiteName} cursor implementation`, () => {
+        describe("extract roundtrip", () => {
+            for (const [name, data] of cursorTestCases) {
+                it(`${name}: ${JSON.stringify(data)}`, () => {
+                    const cursor = factory(data);
+                    const clone = dataFromCursor(cursor);
+                    assert.deepEqual(clone, data);
+                    // Check objects are actually json compatible
+                    const text = JSON.stringify(clone);
+                    const parsed = JSON.parse(text);
+                    assert.deepEqual(parsed, data);
+                });
+            }
+        });
+
+        it("up from root", () => {
+            const cursor = factory({ type: brand("Foo") });
+            assert.throws(() => cursor.exitNode());
+        });
+
+        describe("keys", () => {
+            const getFieldKey = (cursor: ITreeCursor) => cursor.getFieldKey();
+            const getKeysAsSet = (cursor: ITreeCursor) => new Set(mapCursorFields(cursor, getFieldKey));
+
+            it("object", () => {
+                assertSetsEqual(getKeysAsSet(factory({ type: brand("Foo") })), new Set([]));
+                assertSetsEqual(getKeysAsSet(factory({
+                    type: brand("Foo"),
+                    fields: { x: [{ type: brand("Bar") }] },
+                })), new Set([brand("x")]));
+                assertSetsEqual(
+                    getKeysAsSet(factory({
+                        type: brand("Foo"),
+                        fields: { x: [{ type: brand("Bar") }], test: [{ type: brand("Bar"), value: 6 }] },
+                    })),
+                    new Set([brand("x"), brand("test")]),
+                );
+            });
+
+            it("array", () => {
+                // TODO: should empty arrays report this key?
+                assertSetsEqual(
+                    getKeysAsSet(factory({ type: brand("Foo"), fields: { [EmptyKey]: [] } })),
+                    new Set([EmptyKey]),
+                );
+                assertSetsEqual(
+                    getKeysAsSet(factory({
+                        type: brand("Foo"),
+                        fields: { [EmptyKey]: [{ type: brand("Bar"), value: 0 }] },
+                    })),
+                    new Set([EmptyKey]),
+                );
+                assertSetsEqual(
+                    getKeysAsSet(factory({
+                        type: brand("Foo"),
+                        fields: { [EmptyKey]: [{ type: brand("Bar"), value: "test" }, { type: brand("Bar") }] },
+                    })),
+                    new Set([EmptyKey]),
+                );
+            });
+
+            it("string", () => {
+                assertSetsEqual(getKeysAsSet(factory({ type: brand("Foo"), value: "" })), new Set([]));
```

Comment:
I believe assert.deepEqual handles sets, and will produce much nicer assert messages showing diffs than this helper.

Also, `new Set()` does not need and empty array.